demonstrations/tutorial_shor_catalyst.py

```diff
@@ -0,0 +1,1150 @@
+
+                if q != 1:
+                    p = N // q
+
+    return p, q
+
+
+######################################################################
+# To run the algorithm, we must randomly choose ``a``. In principle, we could
+# randomly generate ``a`` within the function above. However, this cannot
+# currently be QJITted. Passing ``n_bits`` as a static argument is also to work
+# around some QJIT limitations. For now, let's see if our algorithm can
+# successfully factor.
+
+from jax import random
+
+
+def factor_with_shor(N, n_shots=100):
+    key = random.PRNGKey(123456789)
+    key, subkey = random.split(key)
+
+    a_choices = jnp.array(list(range(2, N - 1)))
+    a = random.choice(subkey, a_choices)
+
+    while jnp.gcd(a, N) == 1:
+        key, subkey = random.split(key)
+        a = random.choice(subkey, a_choices)
+
+    # The number of bits of N determines the size of the registers.
+    n_bits = int(jnp.ceil(jnp.log2(N)))
+
+    p, q = 0, 0
+
+    # Get the success probabilities
+    num_success = 0
+    for _ in range(n_shots):
+        candidate_p, candidate_q = shors_algorithm(N, a, n_bits)
+        if candidate_p * candidate_q == N:
+            p, q = candidate_p, candidate_q
+            num_success += 1
+
+    return p, q, num_success / n_shots
+
+
+N = 15
+p, q, success_prob = factor_with_shor(N)
+print(f"N = {p} x {q}. Success probability is {success_prob}")
+
+######################################################################
+# Performance and validation
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# We will now verify that QJIT compilation is happening properly. We will run
+# the algorithm for different ``N`` with the same bit width, and different
+# values of ``a``. We expect the first execution, for the very first ``N`` and
+# ``a``, to take longer than the rest.
+
+import time
+import matplotlib.pyplot as plt
+
+# Some 6-bit numbers
+N_values = [33, 39, 51, 55, 57]
+n_bits = int(jnp.ceil(jnp.log2(N_values[0])))
+
+num_a = 3
+
+execution_times = []
+
+key = random.PRNGKey(1010101)
+
+for N in N_values:
+    a_choices = jnp.array(list(range(2, N - 1)))
+    unique_a = []
+
+    while len(unique_a) != num_a:
+        key, subkey = random.split(key)
+        a = random.choice(subkey, a_choices)
+        if jnp.gcd(a, N) == 1:
+            continue
+        unique_a.append(a)
+
+    for a in unique_a:
+        start = time.time()
+        p, q = shors_algorithm(N, a, n_bits)
+        end = time.time()
+        execution_times.append((N, a, end - start))
+
+        start = time.time()
+        p, q = shors_algorithm(N, a, n_bits)
+        end = time.time()
+        execution_times.append((N, a, end - start))
+
+labels = [f"{ex[0]}, {int(ex[1])}" for ex in execution_times][::2]
+times = [ex[2] for ex in execution_times]
+
+plt.scatter(range(len(times)), times, c=[ex[0] for ex in execution_times])
+plt.xticks(range(0, len(times), 2), labels=labels, rotation=80)
+plt.xlabel("N, a")
+plt.ylabel("Runtime (s)");
```

Comment:
I can't figure out how to suppress the text output of the plot. Adding a semicolon does not help. Also, would be great to get some feedback on clarity of the plot and whether it is effective in capturing the main points

Reply:
This plot does show the lack of recompilation for different `a`s and even `N`s, but I think what might be missing is demonstrating why that is valuable to the user 🤔 
What about comparing to the non-jitted version where re-computing and processing new circuits for each `a` takes a lot of time?

Reply:
> I can't figure out how to suppress the text output of the plot. 

Have you tried this version?
```suggestion
plt.ylabel("Runtime (s)")
plt.show()
```